// browserslist-to-es-version/dist/core.js
const ES_VERSIONS_MAP = {
    chrome: [
        51,
        52,
        57,
        64,
        73,
        80,
        85,
        94,
        110,
        119
    ],
    edge: [
        15,
        15,
        15,
        79,
        79,
        80,
        85,
        94,
        110,
        119
    ],
    safari: [
        10,
        10.3,
        11,
        16.4,
        17,
        17,
        17,
        17,
        17,
        17.4
    ],
    firefox: [
        54,
        54,
        54,
        78,
        78,
        80,
        80,
        93,
        115,
        145
    ],
    opera: [
        38,
        39,
        44,
        51,
        60,
        67,
        71,
        80,
        96,
        105
    ],
    samsung: [
        5,
        6.2,
        6.2,
        8.2,
        11.1,
        13,
        14,
        17,
        21,
        25
    ]
};
const aliases = {
    ios_saf: 'safari',
    and_chr: 'chrome',
    and_ff: 'firefox'
};
const renameBrowser = (name)=>aliases[name] || name;
const resolveESVersion = (version, thresholds)=>{
    const index = thresholds.findIndex((threshold)=>version < threshold);
    const defaultVersion = 2024;
    if (-1 === index) return defaultVersion;
    const ES_VERSIONS = [
        2015,
        2016,
        2017,
        2018,
        2019,
        2020,
        2021,
        2022,
        2023,
        2024
    ];
    return ES_VERSIONS[index - 1] ?? 5;
};
function browsersToESVersion(browsers) {
    let esVersion = 2024;
    for (const item of browsers){
        const pairs = item.split(' ');
        if (pairs.length < 2) continue;
        const browser = renameBrowser(pairs[0]);
        const version = Number(pairs[1].split('-')[0]);
        if (Number.isNaN(version)) continue;
        if ('ie' === browser || 'android' === browser && version < 6) {
            esVersion = 5;
            break;
        }
        const versions = ES_VERSIONS_MAP[browser];
        if (!versions) continue;
        const targetVersion = resolveESVersion(version, versions);
        esVersion = Math.min(targetVersion, esVersion);
    }
    return esVersion;
}
export { browsersToESVersion };
