// @rsdoctor/core/dist/inner-plugins/plugins/sourcemapTool.js
import "node:module";
import { Graph } from "@rsdoctor/utils/common";
import { logger, time, timeEnd } from "@rsdoctor/utils/logger";
import { dirname, isAbsolute, resolve } from "path";
import { SourceMapConsumer } from "source-map";
const UNASSIGNED = '[unassigned]';
function bindContextCache(context, namespace, cache, sourceMapDir, sourceRoot) {
    cache = cache || new Map();
    return (source, sourceMapFilenameRegex)=>{
        if (cache.has(source)) return cache.get(source);
        let resolved = UNASSIGNED;
        if (source.startsWith('file://')) resolved = resolve(context, source.replace(/^file:\/\//, ''));
        else if (source.startsWith('webpack://')) if (sourceMapFilenameRegex) {
            const match = source.match(sourceMapFilenameRegex);
            const filePath = match?.[1];
            const hasNamespace = namespace && source.startsWith(`webpack://${namespace}`) || namespace && source.startsWith(`file://${namespace}`);
            const baseDir = hasNamespace ? process.cwd() : context;
            resolved = filePath ? resolve(baseDir, `./${filePath}`) : UNASSIGNED;
        } else resolved = UNASSIGNED;
        else if (isAbsolute(source)) resolved = source;
        else {
            let baseDir = context;
            if (sourceRoot) baseDir = isAbsolute(sourceRoot) ? sourceRoot : sourceMapDir ? resolve(sourceMapDir, sourceRoot) : resolve(context, sourceRoot);
            else if (sourceMapDir) baseDir = sourceMapDir;
            resolved = resolve(baseDir, source);
        }
        cache.set(source, resolved);
        return resolved;
    };
}
async function collectSourceMaps(map, assetLinesCodeList, _compilation, _this, sourceMapFilenameRegex, namespace, skipSources, sourceMapPath) {
    if (map) {
        const consumer = await new SourceMapConsumer(map);
        let sourceMapDir;
        if (sourceMapPath) sourceMapDir = dirname(sourceMapPath);
        else if (_compilation.options.output?.path) sourceMapDir = 'string' == typeof _compilation.options.output.path ? _compilation.options.output.path : void 0;
        const sourceRoot = map.sourceRoot;
        const getRealSourcePath = bindContextCache(_this.sdk.root || process.cwd(), namespace, _this._realSourcePathCache, sourceMapDir, sourceRoot ?? void 0);
        const lineMappings = new Map();
        consumer._absoluteSources = consumer._sources;
        consumer.eachMapping((m)=>{
            if (!lineMappings.has(m.generatedLine)) lineMappings.set(m.generatedLine, []);
            lineMappings.get(m.generatedLine).push(m);
        });
        for (const [lineNum, mappings] of lineMappings.entries()){
            mappings.sort((a, b)=>a.generatedColumn - b.generatedColumn);
            const lineIdx = lineNum - 1;
            if (lineIdx < 0 || lineIdx >= assetLinesCodeList.length) continue;
            const line = assetLinesCodeList[lineIdx];
            for(let i = 0; i < mappings.length; i++){
                const m = mappings[i];
                if (!m.source) continue;
                let realSource = m.source;
                if (realSource.includes('!')) {
                    const parts = realSource.split('!');
                    for(let j = parts.length - 1; j >= 0; j--){
                        const part = parts[j];
                        if (!part) continue;
                        const cleanPart = part.split('??')[0];
                        if (!cleanPart.startsWith('builtin:')) {
                            if (cleanPart.startsWith('/') || cleanPart.includes('\\')) {
                                realSource = cleanPart;
                                break;
                            }
                        }
                    }
                    if (realSource === m.source) {
                        const lastPart = parts[parts.length - 1];
                        realSource = lastPart ? lastPart.split('??')[0] : lastPart;
                    }
                } else if (realSource.includes('??')) realSource = realSource.split('??')[0];
                if (realSource) {
                    if ((realSource.startsWith('webpack://') || realSource.startsWith('file://')) && sourceMapFilenameRegex) realSource = getRealSourcePath(realSource, sourceMapFilenameRegex);
                    else if (!isAbsolute(realSource)) realSource = getRealSourcePath(realSource, sourceMapFilenameRegex);
                }
                if (!realSource || realSource === UNASSIGNED) continue;
                if (skipSources && skipSources.has(realSource)) continue;
                const next = mappings[i + 1];
                const start = m.generatedColumn;
                const end = next ? next.generatedColumn : line.length;
                const codeSegment = line.slice(start, end);
                const prev = _this.sourceMapSets.get(realSource) || '';
                _this.sourceMapSets.set(realSource, prev.concat(codeSegment));
            }
        }
    }
}
async function handleAfterEmitAssets(compilation, _this, sourceMapFilenameRegex, namespace) {
    if ('rspack' in compilation.compiler) {
        _this.sourceMapSets = new Map();
        if (_this.assetsWithoutSourceMap) _this.assetsWithoutSourceMap.clear();
        else _this.assetsWithoutSourceMap = new Set();
        time('ensureModulesChunkGraph.afterEmit.start');
        const assets = [
            ...compilation.getAssets()
        ];
        const skipSources = new Set();
        for (const asset of assets){
            const { assetLinesCodeList, map: mapFromAsset, assetName } = parseAsset(asset, assets, 'js/css');
            let map = mapFromAsset;
            let sourceMapPath;
            if (map) {
                const outputPath = compilation.options.output?.path;
                if (outputPath && 'string' == typeof outputPath && assetName) {
                    const mapFileName = assetName.endsWith('.map') ? assetName : `${assetName}.map`;
                    sourceMapPath = resolve(outputPath, mapFileName);
                }
            } else {
                const sourceMapFile = asset.info.related?.sourceMap;
                const sourceMapFileAssetName = sourceMapFile?.replace(/(\.[^.]+)(\.[^.]+)?$/, '$1');
                if (sourceMapFile) {
                    let sourceMapAsset = assets.find((asset)=>asset.name === sourceMapFile);
                    if (!sourceMapAsset && sourceMapFileAssetName) {
                        const baseNameWithoutHash = Graph.formatAssetName(sourceMapFileAssetName, 'string' == typeof compilation.options.output.filename ? compilation.options.output.filename : void 0);
                        sourceMapAsset = assets.find((asset)=>{
                            const assetBaseName = Graph.formatAssetName(asset.name, 'string' == typeof compilation.options.output.filename ? compilation.options.output.filename : void 0);
                            return assetBaseName.includes(baseNameWithoutHash) && asset.name.endsWith('.map');
                        });
                    }
                    if (sourceMapAsset) {
                        map = JSON.parse(sourceMapAsset.source.source().toString());
                        const outputPath = compilation.options.output?.path;
                        if (outputPath && 'string' == typeof outputPath) sourceMapPath = resolve(outputPath, sourceMapAsset.name);
                    }
                } else {
                    if (assetName && 'string' == typeof assetName && (assetName.endsWith('.js') || assetName.endsWith('.bundle') || assetName.endsWith('.css'))) _this.assetsWithoutSourceMap.add(assetName);
                    continue;
                }
            }
            try {
                await collectSourceMaps(map, assetLinesCodeList, compilation, _this, sourceMapFilenameRegex, namespace, skipSources, sourceMapPath);
            } catch (e) {
                logger.debug(e);
            }
            _this.sourceMapSets.forEach((_value, key)=>{
                if (!skipSources.has(key)) skipSources.add(key);
            });
        }
        timeEnd('ensureModulesChunkGraph.afterEmit.start');
    }
}
async function handleEmitAssets(options) {
    const { compilation, pluginInstance, sourceMapFilenameRegex, namespace } = options;
    if (!('rspack' in compilation.compiler)) {
        pluginInstance.sourceMapSets = new Map();
        time('ensureModulesChunkGraph.afterEmit.start');
        const assets = Object.values(compilation.getAssets()).map((asset, index)=>({
                name: index.toString(),
                source: asset,
                info: asset.info || {}
            }));
        for (const asset of assets){
            const { assetLinesCodeList, map, assetName } = parseAsset(asset, assets, 'map');
            if (!map) continue;
            let sourceMapPath;
            const outputPath = compilation.options.output?.path;
            if (outputPath && 'string' == typeof outputPath && 'string' == typeof assetName) {
                const mapFileName = assetName.endsWith('.map') ? assetName : `${assetName}.map`;
                sourceMapPath = resolve(outputPath, mapFileName);
            }
            try {
                await collectSourceMaps(map, assetLinesCodeList, compilation, pluginInstance, sourceMapFilenameRegex, namespace, void 0, sourceMapPath);
            } catch (e) {
                logger.debug(e);
            }
        }
        timeEnd('ensureModulesChunkGraph.afterEmit.start');
    }
}
function parseAsset(asset, assets, type) {
    const assetName = asset.source?.name || asset.name;
    let assetContent = '';
    let assetLinesCodeList = [];
    let map = null;
    try {
        if ('map' === type && 'string' == typeof assetName && assetName.endsWith('.map') && !asset.name.includes('d.ts')) {
            assetContent = asset.source?.source?.source?.() || '';
            if (!assetContent) {
                logger.debug(`Failed to get source content for asset: ${assetName}`);
                return {
                    assetName,
                    assetContent: '',
                    assetLinesCodeList: [],
                    map: null
                };
            }
            map = JSON.parse(assetContent);
            if (!map || !map.file) {
                logger.debug(`Failed to get source map file for asset: ${assetName}`);
                return {
                    assetName,
                    assetContent,
                    assetLinesCodeList: [],
                    map: null
                };
            }
            const bundledAsset = assets.find((asset2)=>asset2.source?.name === map?.file);
            const bundledCode = bundledAsset?.source?.source?.source?.() || '';
            if (!bundledCode) {
                logger.debug(`Failed to get bundled code for asset: ${map.file}`);
                return {
                    assetName,
                    assetContent,
                    assetLinesCodeList: [],
                    map
                };
            }
            assetLinesCodeList = bundledCode.split(/\r?\n/);
        } else if ('js/css' === type && (assetName.endsWith('.js') || assetName.endsWith('.bundle') || assetName.endsWith('.css'))) {
            assetContent = asset.source?.source?.() || '';
            assetLinesCodeList = assetContent.split(/\r?\n/);
            map = asset.source?.sourceAndMap?.()?.map || null;
        }
    } catch (error) {
        logger.debug(`Error parsing asset ${assetName}:`, error);
        return {
            assetName,
            assetContent: '',
            assetLinesCodeList: [],
            map: null
        };
    }
    return {
        assetName,
        assetContent,
        assetLinesCodeList,
        map
    };
}
export { UNASSIGNED, bindContextCache, collectSourceMaps, handleAfterEmitAssets, handleEmitAssets };
