// @rsdoctor/core/dist/rules/rules/module-mixed-chunks/index.js
import "node:module";
import { Linter } from "@rsdoctor/types";
import { defineRule } from "../../rule.js";
const title = 'module-mixed-chunks';
const rule = defineRule(()=>({
        meta: {
            code: 'E1006',
            title: title,
            category: 'bundle',
            severity: Linter.Severity.Warn,
            defaultConfig: {
                ignore: []
            }
        },
        check ({ moduleGraph, report, ruleConfig }) {
            const modules = moduleGraph.getModules();
            for (const module of modules){
                if (ruleConfig.ignore.some((pattern)=>module.path.includes(pattern))) continue;
                const moduleChunks = module.getChunks();
                if (0 === moduleChunks.length) continue;
                const hasInitial = moduleChunks.some((chunk)=>chunk.initial);
                const hasAsync = moduleChunks.some((chunk)=>!chunk.initial);
                if (hasInitial && hasAsync) {
                    const initialChunks = moduleChunks.filter((chunk)=>chunk.initial);
                    const asyncChunks = moduleChunks.filter((chunk)=>!chunk.initial);
                    const detail = {
                        type: title,
                        module: {
                            id: module.id,
                            path: module.path,
                            webpackId: module.webpackId
                        },
                        initialChunks: initialChunks.map((chunk)=>({
                                id: chunk.id,
                                name: chunk.name
                            })),
                        asyncChunks: asyncChunks.map((chunk)=>({
                                id: chunk.id,
                                name: chunk.name
                            }))
                    };
                    const chunkNames = moduleChunks.map((chunk)=>chunk.name || `chunk-${chunk.id}`).join(', ');
                    const message = `Module "${module.path}" is included in both initial and async chunks: ${chunkNames}. This should be optimized to avoid code duplication.`;
                    report({
                        message,
                        detail
                    });
                }
            }
        }
    }));
export { rule };
