// @rsdoctor/core/dist/rules/rules/default-import-check/index.js
import "node:module";
import { builtinModules } from "module";
import { Linter } from "@rsdoctor/types";
import { getDocument, parser } from "@rsdoctor/utils/ruleUtils";
import { getDefaultImportByRequest, getFixData, getSourceRangeFromTransformedOffset, hasSameLeftInAssignStatement } from "./utils.js";
import { defineRule } from "../../rule.js";
const title = 'default-import-check';
const rule = defineRule(()=>{
    const parserOpt = {
        ecmaVersion: 6
    };
    const exportsDefault = parser.internal.parseExpressionAt('exports.default', 0, parserOpt);
    const moduleExportsDefault = parser.internal.parseExpressionAt('module.exports.default', 0, parserOpt);
    return {
        meta: {
            code: 'E1005',
            title: title,
            category: 'compile',
            severity: Linter.Severity.Warn,
            defaultConfig: {
                ignore: []
            }
        },
        check ({ moduleGraph, report, ruleConfig }) {
            const dependencyWithNode = moduleGraph.getDependencies().filter((dep)=>'dynamic' === dep.meta.exportsType).filter((dep)=>!dep.module.path.includes('node_modules')).filter((dep)=>!ruleConfig.ignore.includes(dep.request) && !builtinModules.includes(dep.request)).filter((dep)=>dep.dependency.meta.hasSetEsModuleStatement).filter((dep)=>dep.dependency.getProgram()).filter((dep)=>!hasSameLeftInAssignStatement(dep.dependency.getProgram(), [
                    exportsDefault,
                    moduleExportsDefault
                ])).map((dependency)=>{
                const { module, request } = dependency;
                const node = module.getProgram() && getDefaultImportByRequest(module.getProgram(), request);
                return node ? {
                    dependency,
                    node
                } : {};
            }).filter((dep)=>dep?.dependency);
            for (const { dependency, node } of dependencyWithNode){
                if (!dependency) continue;
                const document = getSourceRangeFromTransformedOffset(dependency.module, node);
                if (!document) continue;
                const message = 'Do not to use the default import when you import a cjs module.';
                const canFix = !document.isTransformed && !document.path.includes('node_modules');
                if (canFix) {
                    const fixData = getFixData(dependency.module, node, document.range);
                    const detail = {
                        type: 'code-change',
                        file: {
                            path: document.path,
                            actual: document.content,
                            line: document.range.start.line,
                            expected: getDocument(document.content).edit(fixData)
                        }
                    };
                    report({
                        message,
                        document,
                        detail,
                        suggestions: {
                            description: 'Use namespace import instead.',
                            fixData
                        }
                    });
                } else {
                    const detail = {
                        type: 'code-view',
                        file: {
                            path: document.path,
                            content: document.content,
                            ranges: [
                                node.loc
                            ]
                        }
                    };
                    report({
                        message,
                        document,
                        detail
                    });
                }
            }
        }
    };
});
export { rule };
