// @rsdoctor/core/dist/rules/rules/duplicate-package/index.js
import "node:module";
import { Linter } from "@rsdoctor/types";
import { diff, gt } from "semver";
import { getErrorDetail, getErrorMsg } from "./utils.js";
import { defineRule } from "../../rule.js";
import * as __rspack_external__types_js_7eec16d2 from "./types.js";
const title = 'duplicate-package';
const { CheckVersionMap: CheckVersionMap } = __rspack_external__types_js_7eec16d2;
const rule = defineRule(()=>({
        meta: {
            code: 'E1001',
            title: title,
            category: 'bundle',
            severity: Linter.Severity.Warn,
            defaultConfig: {
                checkVersion: 'major',
                ignore: []
            }
        },
        check ({ packageGraph, report, root, ruleConfig }) {
            const checkVersion = CheckVersionMap[ruleConfig.checkVersion];
            const packages = packageGraph.getDuplicatePackages().filter((pkg)=>!ruleConfig.ignore.includes(pkg[0].name)).map((pkgs)=>pkgs.filter((current)=>{
                    const check = pkgs.reduce((ans, pkg)=>ans | CheckVersionMap[diff(current.version, pkg.version) ?? 'null'], 0);
                    return check <= checkVersion;
                }).sort((_packA, _packB)=>gt(_packA.version, _packB.version) ? 1 : -1)).filter((pkgs)=>pkgs.length > 1 && pkgs.filter((pkg)=>pkg.getSize().parsedSize > 0).length > 1);
            for (const pkg of packages){
                const message = getErrorMsg(pkg, root);
                const detail = {
                    type: 'package-relation',
                    packages: pkg.map((item)=>getErrorDetail(item, packageGraph))
                };
                report({
                    message,
                    detail
                });
            }
        }
    }));
export { rule };
