// rslog/dist/index.js
import node_util from "node:util";
function checkNodeVersion() {
    const { versions } = process;
    if ("styleText" in node_util || !versions.node || versions.bun || versions.deno) return;
    throw new Error(`Unsupported Node.js version: "${process.versions.node || 'unknown'}". Expected Node.js >= 20.`);
}
checkNodeVersion();
const createStyler = (style)=>(text)=>node_util.styleText(style, String(text));
const color = {
    dim: createStyler('dim'),
    red: createStyler('red'),
    bold: createStyler('bold'),
    blue: createStyler('blue'),
    cyan: createStyler('cyan'),
    gray: createStyler('gray'),
    black: createStyler('black'),
    green: createStyler('green'),
    white: createStyler('white'),
    reset: createStyler('reset'),
    yellow: createStyler('yellow'),
    magenta: createStyler('magenta'),
    underline: createStyler('underline'),
    strikethrough: createStyler('strikethrough')
};
const supportsTrueColor = true === process.stdout.isTTY && process.stdout.hasColors(2 ** 24, process.env);
const boldMint = (text)=>supportsTrueColor ? `\x1b[1;38;2;132;225;199m${text}\x1b[39;22m` : color.bold(color.cyan(text));
const LOG_LEVEL = {
    silent: -1,
    error: 0,
    warn: 1,
    info: 2,
    log: 2,
    verbose: 3
};
const createLogType = (label, level, style)=>{
    let formattedLabel;
    return {
        get label () {
            return formattedLabel ??= color.bold(style(label.padEnd(7)));
        },
        level
    };
};
const LOG_TYPES = {
    error: createLogType('error', 'error', color.red),
    warn: createLogType('warn', 'warn', color.yellow),
    info: createLogType('info', 'info', color.cyan),
    start: createLogType('start', 'info', color.cyan),
    ready: createLogType('ready', 'info', color.green),
    success: createLogType('success', 'info', color.green),
    log: {
        level: 'info'
    },
    debug: createLogType('debug', 'verbose', color.magenta)
};
const ANSI_REGEXP = new RegExp(`${String.fromCharCode(27)}\\[[0-9;]*m`, 'g');
const stripAnsi = (text)=>text.replace(ANSI_REGEXP, '');
const errorStackRegExp = /at [^\r\n]{0,200}(?::\d+:\d+[\s)]*|\(<anonymous>\)|\(index\s\d+\))$/;
const isErrorStackMessage = (message)=>{
    const stripped = stripAnsi(message);
    return errorStackRegExp.test(stripped);
};
const normalizeErrorMessage = (err)=>{
    if (err.stack) {
        const [rawName, ...rest] = err.stack.split('\n');
        const name = rawName.startsWith('Error: ') ? rawName.slice(7) : rawName;
        return `${name}\n${color.gray(rest.join('\n'))}`;
    }
    return err.message;
};
const createLogger = (options = {})=>{
    const { level = 'info', prefix, console = globalThis.console, alignMultiline } = options;
    let maxLevel = level;
    const log = (type, message, ...args)=>{
        const logType = LOG_TYPES[type];
        const { level } = logType;
        if (LOG_LEVEL[level] > LOG_LEVEL[maxLevel]) return;
        if (null == message) return console.log();
        const label = 'label' in logType ? logType.label : '';
        const header = label ? prefix ? `${label} ${prefix}` : label : prefix;
        let text = '';
        const isErrorObject = message instanceof Error;
        if (isErrorObject) {
            text += normalizeErrorMessage(message);
            const { cause } = message;
            if (cause) {
                text += color.yellow('\n  [cause]: ');
                text += cause instanceof Error ? normalizeErrorMessage(cause) : String(cause);
            }
        } else if ('error' === level && 'string' == typeof message) {
            const lines = message.split('\n');
            text = lines.map((line)=>isErrorStackMessage(line) ? color.gray(line) : line).join('\n');
        } else text = `${message}`;
        if (alignMultiline && header && !isErrorObject && text.includes('\n')) {
            const indent = ' '.repeat(stripAnsi(header).length + 1);
            const skipStack = 'error' === level;
            text = text.split('\n').map((line, i)=>0 === i || '' === line || skipStack && isErrorStackMessage(line) ? line : indent + line).join('\n');
        }
        const method = 'error' === level || 'warn' === level ? level : 'log';
        console[method](header ? `${header} ${text}` : text, ...args);
    };
    const logger = {
        greet: (message)=>log('log', boldMint(message))
    };
    Object.keys(LOG_TYPES).forEach((key)=>{
        logger[key] = (...args)=>log(key, ...args);
    });
    Object.defineProperty(logger, 'level', {
        get: ()=>maxLevel,
        set (val) {
            maxLevel = val;
        }
    });
    Object.defineProperty(logger, 'options', {
        get: ()=>({
                ...options
            })
    });
    logger.override = (customLogger)=>{
        Object.assign(logger, customLogger);
    };
    return logger;
};
const src_logger = createLogger();
export { color, createLogger, src_logger as logger };
