// @rsdoctor/core/dist/rules/rules/loader-performance-optimization/index.js
import "node:module";
import { Linter } from "@rsdoctor/types";
import { Loader, Time } from "@rsdoctor/utils/common";
import { defineRule } from "../../rule.js";
import { match } from "./utils.js";
const title = 'loader-performance-optimization';
const rule = defineRule(()=>({
        meta: {
            code: 'E1003',
            title: title,
            category: 'compile',
            severity: Linter.Severity.Warn,
            defaultConfig: {
                ignore: [],
                threshold: 5000,
                extensions: [
                    'js',
                    'css',
                    'jpg',
                    'jpeg',
                    'png',
                    'gif',
                    'webp',
                    'svg'
                ]
            }
        },
        check ({ loader, configs, root, ruleConfig, report }) {
            const { extensions, ignore = [], threshold = 5000 } = ruleConfig;
            if (!extensions || extensions?.length === 0) return;
            if (0 === loader.length) return;
            const { config } = configs.find((e)=>'webpack' === e.name) || {
                config: void 0
            };
            if (!config) return;
            const cwd = config.context || root;
            const nodeModulesPathRegexp = /\/node_modules\//;
            const resultMap = new Map();
            const loaders = loader.flatMap((el)=>el.loaders);
            for (const item of loader){
                const { path, ext } = item.resource;
                if (match(ext, extensions)) {
                    if (nodeModulesPathRegexp.test(path) || !path.startsWith(cwd)) item.loaders.forEach((el)=>{
                        if (match(el.loader, ignore)) return;
                        const costs = Loader.getLoaderCosts(el, loaders);
                        const v = {
                            ...el,
                            __resource__: item.resource,
                            __costs__: costs
                        };
                        if (resultMap.has(el.loader)) resultMap.get(el.loader).push(v);
                        else resultMap.set(el.loader, [
                            v
                        ]);
                    });
                }
            }
            const sum = (arr)=>{
                const v1 = arr.reduce((t, c)=>t + c.__costs__, 0);
                const v2 = Math.max.apply(null, arr.map((e)=>e.endAt)) - Math.min.apply(arr.map((e)=>e.startAt));
                if (v2 >= v1 || v2 === -1 / 0) return v1;
                return v2;
            };
            for (const [loaderName, v] of resultMap.entries()){
                const nodeModulesFiles = v.filter((e)=>nodeModulesPathRegexp.test(e.__resource__.path));
                const nodeModulesFilesCosts = sum(nodeModulesFiles);
                const outofCwdFiles = v.filter((e)=>!e.__resource__.path.startsWith(cwd));
                const outofCwdFilesCosts = sum(outofCwdFiles);
                const msg = [
                    nodeModulesFilesCosts >= threshold && `${nodeModulesFiles.length} node_modules files (costs: ${Time.formatCosts(nodeModulesFilesCosts)})`,
                    outofCwdFilesCosts >= threshold && `${outofCwdFiles.length} outside project files (costs: ${Time.formatCosts(outofCwdFilesCosts)})`
                ].filter(Boolean).join(' and ');
                if (msg) {
                    const message = `${loaderName} has processed ${msg}.`;
                    report({
                        message,
                        detail: {
                            description: message,
                            type: 'link'
                        }
                    });
                }
            }
        }
    }));
export { rule };
