// @rsdoctor/core/dist/rules/rules/esm-resolved-to-cjs/index.js
import "node:module";
import fs from "fs";
import path from "path";
import { Linter, SDK } from "@rsdoctor/types";
import { defineRule } from "../../rule.js";
const title = 'esm-resolved-to-cjs';
function normalizePathForCompare(filePath) {
    const queryIndex = filePath.indexOf('?');
    const hashIndex = filePath.indexOf('#');
    const splitIndex = [
        queryIndex,
        hashIndex
    ].filter((index)=>index >= 0).reduce((min, index)=>Math.min(min, index), 1 / 0);
    const cleanPath = splitIndex === 1 / 0 ? filePath : filePath.slice(0, splitIndex);
    return path.normalize(cleanPath);
}
function isDefinitelyEsmFile(filePath) {
    return '.mjs' === path.extname(normalizePathForCompare(filePath));
}
function isDeclaredEsmEntry(filePath, esmEntries, toRealPath) {
    const realPath = toRealPath(filePath);
    return esmEntries.some((esmEntry)=>realPath === toRealPath(esmEntry));
}
function isRecord(value) {
    return null !== value && 'object' == typeof value && !Array.isArray(value);
}
function substituteExportPattern(target, wildcard) {
    return target.replace(/\*/g, wildcard);
}
function toPackagePath(pkgRoot, target, wildcard) {
    return path.resolve(pkgRoot, substituteExportPattern(target, wildcard));
}
function collectStringTargets(value, wildcard) {
    if ('string' == typeof value) return [
        substituteExportPattern(value, wildcard)
    ];
    if (Array.isArray(value)) return value.flatMap((item)=>collectStringTargets(item, wildcard));
    if (!isRecord(value)) return [];
    return Object.values(value).flatMap((child)=>collectStringTargets(child, wildcard));
}
function collectEsmTargets(exportEntry, pkgRoot, wildcard) {
    const entries = new Set();
    const addTargets = (value)=>{
        for (const target of collectStringTargets(value, wildcard))entries.add(toPackagePath(pkgRoot, target, ''));
    };
    const visit = (value)=>{
        if ('string' == typeof value || Array.isArray(value)) return void addTargets(value);
        if (!isRecord(value)) return;
        addTargets(value['import']);
        addTargets(value['module']);
        for (const child of Object.values(value))if (isRecord(child) || Array.isArray(child)) visit(child);
    };
    visit(exportEntry);
    return entries;
}
function hasSubpathExportKeys(exports) {
    return Object.keys(exports).some((key)=>'.' === key || key.startsWith('./'));
}
function matchPatternExportKey(patternKey, exportKey) {
    const wildcardIndex = patternKey.indexOf('*');
    if (wildcardIndex < 0) return null;
    const prefix = patternKey.slice(0, wildcardIndex);
    const suffix = patternKey.slice(wildcardIndex + 1);
    if (!exportKey.startsWith(prefix) || !exportKey.endsWith(suffix)) return null;
    return exportKey.slice(prefix.length, exportKey.length - suffix.length);
}
function isPatternExportKey(key) {
    return key.indexOf('*') === key.lastIndexOf('*') && key.includes('*');
}
function patternKeyCompare(keyA, keyB) {
    const baseLengthA = keyA.includes('*') ? keyA.indexOf('*') + 1 : keyA.length;
    const baseLengthB = keyB.includes('*') ? keyB.indexOf('*') + 1 : keyB.length;
    if (baseLengthA > baseLengthB) return -1;
    if (baseLengthB > baseLengthA) return 1;
    if (!keyA.includes('*')) return 1;
    if (!keyB.includes('*')) return -1;
    if (keyA.length > keyB.length) return -1;
    if (keyB.length > keyA.length) return 1;
    return 0;
}
function resolveExportEntry(exports, exportKey) {
    if (Object.prototype.hasOwnProperty.call(exports, exportKey)) return {
        value: exports[exportKey],
        wildcard: ''
    };
    const patternEntries = Object.entries(exports).filter(([key])=>isPatternExportKey(key)).sort(([keyA], [keyB])=>patternKeyCompare(keyA, keyB));
    for (const [key, value] of patternEntries){
        const wildcard = matchPatternExportKey(key, exportKey);
        if (null !== wildcard) return {
            value,
            wildcard
        };
    }
    return null;
}
function resolveExportEntryForKey(exportsField, exportKey) {
    if (!isRecord(exportsField)) return null;
    if (!hasSubpathExportKeys(exportsField)) return '.' === exportKey ? {
        value: exportsField,
        wildcard: ''
    } : null;
    return resolveExportEntry(exportsField, exportKey);
}
function extractEsmEntries(pkgJson, pkgRoot, exportKey) {
    const { exports: exportsField, module: moduleField } = pkgJson;
    const resolvedExport = resolveExportEntryForKey(exportsField, exportKey);
    const entries = resolvedExport ? collectEsmTargets(resolvedExport.value, pkgRoot, resolvedExport.wildcard) : new Set();
    if ('.' === exportKey && 'string' == typeof moduleField) entries.add(path.resolve(pkgRoot, moduleField));
    return entries;
}
function getExportKeyFromRequest(packageName, request) {
    if (request === packageName) return '.';
    if (request.startsWith(`${packageName}/`)) return `.${request.slice(packageName.length)}`;
    return null;
}
const rule = defineRule(()=>({
        meta: {
            code: 'E1009',
            title: title,
            category: 'bundle',
            severity: Linter.Severity.Warn,
            defaultConfig: {
                ignore: []
            }
        },
        check ({ moduleGraph, packageGraph, report, ruleConfig }) {
            const pkgJsonCache = new Map();
            const realPathCache = new Map();
            const readPkgJson = (pkgRoot)=>{
                if (pkgJsonCache.has(pkgRoot)) return pkgJsonCache.get(pkgRoot);
                try {
                    const raw = fs.readFileSync(path.join(pkgRoot, 'package.json'), 'utf-8');
                    const parsed = JSON.parse(raw);
                    pkgJsonCache.set(pkgRoot, parsed);
                    return parsed;
                } catch  {
                    pkgJsonCache.set(pkgRoot, null);
                    return null;
                }
            };
            const toRealPath = (filePath)=>{
                const normalized = normalizePathForCompare(filePath);
                if (realPathCache.has(normalized)) return realPathCache.get(normalized);
                try {
                    const realPath = 'function' == typeof fs.realpathSync.native ? fs.realpathSync.native(normalized) : fs.realpathSync(normalized);
                    const normalizedRealPath = path.normalize(realPath);
                    realPathCache.set(normalized, normalizedRealPath);
                    return normalizedRealPath;
                } catch  {
                    realPathCache.set(normalized, normalized);
                    return normalized;
                }
            };
            const groups = new Map();
            const inNodeModules = /[/\\]node_modules[/\\]/;
            for (const dep of moduleGraph.getDependencies()){
                if (dep.kind !== SDK.DependencyKind.ImportStatement || inNodeModules.test(dep.module.path) || !inNodeModules.test(dep.dependency.path) || dep.dependency.meta.strictHarmonyModule || ruleConfig.ignore.some((p)=>dep.request.includes(p))) continue;
                const pkg = packageGraph.getPackageByModule(dep.dependency);
                if (!pkg?.root) continue;
                const pkgJson = readPkgJson(pkg.root);
                if (ruleConfig.ignore.some((p)=>pkg.name.includes(p)) || !pkgJson) continue;
                const exportKey = getExportKeyFromRequest(pkg.name, dep.request);
                if (!exportKey) continue;
                const esmEntries = extractEsmEntries(pkgJson, pkg.root, exportKey);
                if (0 === esmEntries.size) continue;
                const esmEntryList = Array.from(esmEntries);
                const resolvedModuleRealPath = toRealPath(dep.dependency.path);
                if (isDefinitelyEsmFile(resolvedModuleRealPath)) continue;
                if (isDeclaredEsmEntry(resolvedModuleRealPath, esmEntryList, toRealPath)) continue;
                const groupKey = `${pkg.name}::${resolvedModuleRealPath}`;
                const issuer = {
                    id: dep.module.id,
                    path: dep.module.path,
                    request: dep.request
                };
                if (groups.has(groupKey)) groups.get(groupKey).issuers.push(issuer);
                else groups.set(groupKey, {
                    packageName: pkg.name,
                    packageVersion: pkg.version,
                    esmEntry: esmEntryList[0],
                    resolvedModule: {
                        id: dep.dependency.id,
                        path: dep.dependency.path,
                        webpackId: dep.dependency.webpackId
                    },
                    issuers: [
                        issuer
                    ]
                });
            }
            for (const group of groups.values()){
                const detail = {
                    type: title,
                    packageName: group.packageName,
                    packageVersion: group.packageVersion,
                    esmEntry: group.esmEntry,
                    resolvedModule: group.resolvedModule,
                    issuers: group.issuers
                };
                const message = `"${group.packageName}" declares an ESM entry ("${group.esmEntry}") but was resolved to CJS by ${group.issuers.length} ESM import(s). Check your bundler's \`resolve.mainFields\` or \`resolve.conditionNames\` config.`;
                report({
                    message,
                    detail
                });
            }
        }
    }));
export { rule };
